src/main/java/org/mule/extension/db/internal/source/RowListener.java

```diff
@@ -0,0 +1,215 @@
+ * will be used to filter the contents of the next poll, so that only rows with a greater watermark value are returned. If an
+ * id column is provided, this component will automatically make sure that the same row is not picked twice by concurrent polls
+ *
+ * @since 1.1
+ */
+@MetadataScope(outputResolver = RowListenerMetadataResolver.class)
+@Alias("listener")
+public class RowListener extends PollingSource<Map<String, Object>, Void> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RowListener.class);
+
+  /**
+   * The name of the table to select from
+   */
+  @Parameter
+  @MetadataKeyId
+  private String table;
+
+  /**
+   * The name of the column to use for watermark
+   */
+  @Parameter
+  @Optional
+  private String watermarkColumn;
+
+  /**
+   * The name of the column to consider as row ID. If provided, this component will make sure that the same row is not
+   * processed twice by concurrent polls.
+   */
+  @Parameter
+  @Optional
+  private String idColumn;
+
+  @ParameterGroup(name = ADVANCED)
+  @NullSafe
+  private QuerySettings settings;
+
+  @Config
+  private DbConnector config;
+
+  @Connection
+  private ConnectionProvider<DbConnection> connectionProvider;
+
+  private final QueryResolver<ParameterizedStatementDefinition> queryResolver = new ParameterizedQueryResolver<>();
+  private ItemHandler idHandler;
+  private ItemHandler watermarkHandler;
+
+  @Override
+  protected void doStart() throws MuleException {
+    if (idColumn != null) {
+      idHandler = (item, row) -> {
+        Object id = row.get(idColumn);
+        if (id != null) {
+          item.setId(id.toString());
+        }
+      };
+    } else {
+      idHandler = new NullItemHandler();
+    }
+
+    if (watermarkColumn != null) {
+      watermarkHandler = (item, row) -> {
+        Object watermark = row.get(watermarkColumn);
+        if (watermark == null) {
+          if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(format(
+                                "A null watermark value was obtained for row %s. Watermark value won't be updated for this row",
+                                row));
+          }
+
+          return;
+        }
+
+        if (!(watermark instanceof Serializable)) {
+          LOGGER.error(format("Watermark values need to be serializable, but a value of type %s was found instead for row %s",
+                              watermark.getClass().getName(), row));
+        }
+
+        item.setWatermark((Serializable) watermark);
+      };
+    } else {
+      watermarkHandler = new NullItemHandler();
+    }
+  }
+
+  @Override
+  protected void doStop() {
+
+  }
+
+  @Override
+  public void poll(PollContext<Map<String, Object>, Void> pollContext) {
+    if (pollContext.isSourceStopping()) {
+      return;
+    }
+
+    DbConnection connection;
+    try {
+      connection = connectionProvider.connect();
```

Comment:
Why can't you just ask for a Connection in the poll method signature?.

Reply:
this was a design decision since not all use cases will handle this in the same way